### PR TITLE
Siege Cannon

### PR DESCRIPTION
**Pull request by GreatBigFailure**
Siege Cannon is a new makeshift weapon intended as a (significantly) upgraded version of the current cannon. Excellent cannon sprite by Blithering.

siegecannon

### Construction

Siege Cannons are made by assembling a normal cannon: gun barrel -> wheelchair -> weld, but do not add the igniter. Instead use a barricade kit on the assembly for a wooden frame, then 10 sheets of plasteel to re-enforce the barrel. Now you have your own Siege Cannon. Add welder fuel to it like a normal cannon, though it accepts 20 rather than 10.

### Damage

Siege Cannons differ from most "Just throw something really hard" style weapons, like IEDs and normal cannons, because they prefer larger projectiles over small. That means glass shards are no longer king when it comes to damage. Toolboxes are going to break some bones and a few rarely mentioned items might even do worse.
The counterplay is obviously to not stand in front of the cannon.

### Ammunition

Siege Cannons can be loaded with anything a normal cannon can and then some. They even accept humans if you want to be a death defying stuntman, or a particularly cruel warden.

nowloadingclown

If your munition of choice is, for some reason, humans then an obese one will hit even harder. With that said what you should really be loading is the new **cannonball**, made with 20 sheets of metal.

### Lay Siege

The real point of the Siege Cannon.
When a cannonball is fired from a Siege Cannon it becomes a wall damaging, computer destroying, machine breaking, HoS stunning projectile. Now you can siege the malf's core in the name of the king. 

harkandverily

Mind you it may take a few shots to tear down tougher walls, and cannonballs are too bulky for one hand to carry.
Cannonballs don't just break walls, mind you. They'll also smash through the brig's windows, computers, and anything else standing between you and your plunder, yar har.

YAAAAAR

Cannonballs can only be carried with two hands and don't have their sieging properties when thrown/fired from anything but a Siege Cannon.

Sorry.

:cl:
 * rscadd: Cannonballs can now be constructed with 20 sheets of metal.
 * rscadd: New siege cannons facilitate knight, pirate, and greyshit RP.  They also destroy walls when firing cannonballs.

### CODE REVIEW

**Comment by SageOfLight** [comment]
CAN THE MANNONS.

**Comment by BomberBro** [comment]
I like it, isn't the cannonball sprite a little big though?

**Comment by GreatBigFailure** [comment]
> I like it, isn't the cannonball sprite a little big though?

Excellent cannon sprite by Blithering
Awful cannonball codersprite by me

**Comment by Dacendeth** [comment]
the cannonball sprite is way to big.

**Comment by APCUnlocked** [comment]
Sounds great, can't wait for the historical revolutionaries to re-enact the storming of Bastille

**Comment by Eneocho** [comment]
It wrecks walls, can it wreck doors or blast doors?

**Comment by GreatBigFailure** [comment]
> It wrecks walls, can it wreck doors or blast doors?

It takes much longer but yes it can destroy doors.
It mimics an explosion on most things. For airlocks it picks an explosion severity of 2 or 3. 3 just makes the airlock spark, 2 has a 25% chance of destroying it. So essentially a 12.5% chance of destroying an airlock per shot. Blast doors it tears through pretty quickly which is intentional. 
It takes a decent bit of time to get through things. Faster than disassembling them for sure and at a range but it's not one shot = no wall. It disassembles in stages and based on RNG. So wall -> partial wall -> girder -> nothing

**Comment by Blithering** [comment]
I can't wait for the first malf AI to get taken out by a cargo train broadside barrage.

**Comment by GreatBigFailure** [comment]
Little update:

- Limited loading items to weight class large and below. This is practically only due to mech parts being huge and therefor an instant-gib. I thought you couldn't even pick up anything bigger than large.
- Added more feedback messages when doing things
- No longer restricted filling the cannon to beaker-type containers
- You can now wrench it in place. You can still use the right click verb to rotate it while wrenched, this is intentional

Regarding the first point, I tested practically every large item with a decent throw_force I could find and they should be fine. Nothing hit harder than _redacted_, king of thrown weapons.
owmyblood

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+		loadedItem.forceMove(loc)
+		loadedItem = null
+	if(loadedMob)
+		loadedMob.forceMove(loc)
+		loadedMob = null
+
+/obj/structure/siege_cannon/AltClick(var/mob/user)
+	if(user.incapacitated() || !in_range(user, src) || user.loc == src)
+		return
+	to_chat(user,"<span class='notice'>You unload the [src].</span>" )
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
	to_chat(user,"<span class='notice'>You unload \the [src].</span>" )
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	unloadCannon()
+
+/obj/structure/siege_cannon/attackby(obj/item/W as obj, mob/user as mob)
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]: `as mob` and `as objs` are not needed.

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+	if(istype(W, /obj/item/weapon/reagent_containers))
+		fillCannon(W, user)
+		return 1
+	if(W.is_wrench())
+		wrenchAnchor(user, W, 5)	//Half a second to wrench. Being able to turn it via verb while anchored is intentional.
+	else
+		loadCannon(W, user)
+
+/obj/structure/siege_cannon/attack_hand(mob/user)
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm:40]: `attack_hand` is called for every kind of mob, including ghosts and dead people. While there is sanity for a human missing his hands or being restrained, there is *no* sanity for a dead human or a ghost. Add a stat check.

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+		return
+	for(var/datum/reagent/R in G.reagents.reagent_list)
+		if(R.id != FUEL)
+			to_chat(user,"<span class='warning'>The [src] can't accept that as fuel.</span>" )
+		else
+			var/tF = clamp(G.amount_per_transfer_from_this, 0, maxFuel - wFuel)
+			G.reagents.remove_reagent(FUEL, tF)
+			wFuel += tF
+			if(wFuel == maxFuel)
+				to_chat(user,"<span class='notice'>You completely fill the [src] with [tF] units of fuel.</span>" )
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
				to_chat(user,"<span class='notice'>You completely fill \the [src] with [tF] units of fuel.</span>" )
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+			else
+				to_chat(user,"<span class='notice'>You add [tF] units of fuel to \the [src].</span>" )
+
+/obj/structure/siege_cannon/proc/loadCannon(var/obj/item/cAmmo, var/mob/user)
+	if(loadedItem || loadedMob)
+		return
+	if(cAmmo.w_class > maxSize)
+		to_chat(user,"<span class='warning'>The [cAmmo] is too large to fit in the [src].</span>")
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
		to_chat(user,"<span class='warning'>The [cAmmo] is too large to fit in \the [src].</span>")
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		return
+	if(user.drop_item(cAmmo, src))
+		loadedItem = cAmmo
+		to_chat(user,"<span class='notice'>You load \the [cAmmo] into the [src].</span>" )
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
		to_chat(user,"<span class='notice'>You load \the [cAmmo] into \the [src].</span>" )
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+/obj/structure/siege_cannon/MouseDropTo(var/atom/movable/C, mob/user)
+	if(user.incapacitated() || !in_range(user, src) || !in_range(user, C) || C.anchored)	//Copy pasted from chairs because sanity is hard
+		return
+	if(!isliving(C))
+		return
+	if(loadedMob || loadedItem)
+		to_chat(user,"<span class='warning'>The [src] is already full.</span>" )
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
		to_chat(user,"<span class='warning'>\The [src] is already full.</span>" )
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+		return
+	visible_message("<span class='warning'>[user] is stuffing [C] into \the [src].</span>")
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]:
```suggestion
	visible_message("<span class='warning'>\The [user] is stuffing [C] into \the [src].</span>")
```

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+		throw_range = 2
+		throw_speed = 1
+		throwforce = 3
+
+/obj/item/cannonball/throw_impact(atom/hit_atom)
+	..()
+	if(cannonFired && !hit_atom.gcDestroyed)
+		cannonEffect(hit_atom)
+
+/obj/item/cannonball/proc/cannonEffect(var/atom/cTarg)
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm:192]: This is such a crime against OOP that I cannot in good faith support it.
Atoms already have an `hitby` proc which seems like a more natural place to put this.

File: code/modules/projectiles/guns/projectile/constructable/siegecannon.dm [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+			M.spillContents()
+			qdel(M)
+		else if(M.wrenchable())
+			M.state = 0
+			M.anchored = FALSE
+			M.power_change()
+	else
+		spark(M)
+
+/obj/item/cannonball/proc/siegeMob(var/mob/living/L)
```
Comment (ShiftyRail) [inline code/modules/projectiles/guns/projectile/constructable/siegecannon.dm]: Likewise, `mob/hitby` is where this should go.

**Review by Kurfursten — CHANGES_REQUESTED** [review]
Address Shifty's review

**Pull request merged**